<a href="https://colab.research.google.com/github/MavelSterling/Informe_Inteligencia_Artificial/blob/main/notebook_Redes_Neuronales.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# REDES NEURONALES EN DATOS DE SOBREVIVIENTES DEL TITANIC

In [8]:
#----------------------------------------------------------
# Se importan las librerias
#----------------------------------------------------------

import numpy as np
import pandas as pd
import sklearn
import warnings

## 1. Leer el archivo titanic.csv. Tenga en cuenta que los elementos del archivo están separados por ‘;’

In [9]:
#----------------------------------------------------------
# Url de los datos
#----------------------------------------------------------

url= ("https://raw.githubusercontent.com/MavelSterling/Informe_Inteligencia_Artificial/main/titanic.csv")

#----------------------------------------------------------
# Se cargan los datos
#----------------------------------------------------------
datos = pd.read_csv(url, sep=";")
datos


survived                                               Name  pclass  \
0           0                            Braund, Mr. Owen Harris       3   
1           1  Cumings, Mrs. John Bradley (Florence Briggs Th...       1   
2           1                             Heikkinen, Miss. Laina       3   
3           1       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1   
4           0                           Allen, Mr. William Henry       3   
..        ...                                                ...     ...   
886         0                              Montvila, Rev. Juozas       2   
887         1                       Graham, Miss. Margaret Edith       1   
888         0           Johnston, Miss. Catherine Helen "Carrie"       3   
889         1                              Behr, Mr. Karl Howell       1   
890         0                                Dooley, Mr. Patrick       3   

        sex   age  sibsp  parch     fare    who  embark_town alone  
0      male  22.0      1      0   7.2500    man  Southampton    no  
1    female  38.0      1      0  71.2833  woman    Cherbourg    no  
2    female  26.0      0      0   7.9250  woman  Southampton   yes  
3    female  35.0      1      0  53.1000  woman  Southampton    no  
4      male  35.0      0      0   8.0500    man  Southampton   yes  
..      ...   ...    ...    ...      ...    ...          ...   ...  
886    male  27.0      0      0  13.0000    man  Southampton   yes  
887  female  19.0      0      0  30.0000  woman  Southampton   yes  
888  female   NaN      1      2  23.4500  woman  Southampton    no  
889    male  26.0      0      0  30.0000    man    Cherbourg   yes  
890    male  32.0      0      0   7.7500    man   Queenstown   yes  

[891 rows x 11 columns]

In [10]:
#----------------------------------------------------------
# Información de la base de datos
#----------------------------------------------------------

datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   Name         891 non-null    object 
 2   pclass       891 non-null    int64  
 3   sex          891 non-null    object 
 4   age          714 non-null    float64
 5   sibsp        891 non-null    int64  
 6   parch        891 non-null    int64  
 7   fare         891 non-null    float64
 8   who          891 non-null    object 
 9   embark_town  889 non-null    object 
 10  alone        891 non-null    object 
dtypes: float64(2), int64(4), object(5)
memory usage: 76.7+ KB


## 2. Eliminar el atributo Name usando el comando datos.drop(["Name"], axis = 1, inplace = True)

In [11]:
#----------------------------------------------------------
# Eliminar el atributo Name
#----------------------------------------------------------

datos.drop(["Name"], axis = 1, inplace = True)

#----------------------------------------------------------
# Observar los datos sin el atributo Name
#----------------------------------------------------------
datos

survived  pclass     sex   age  sibsp  parch     fare    who  \
0           0       3    male  22.0      1      0   7.2500    man   
1           1       1  female  38.0      1      0  71.2833  woman   
2           1       3  female  26.0      0      0   7.9250  woman   
3           1       1  female  35.0      1      0  53.1000  woman   
4           0       3    male  35.0      0      0   8.0500    man   
..        ...     ...     ...   ...    ...    ...      ...    ...   
886         0       2    male  27.0      0      0  13.0000    man   
887         1       1  female  19.0      0      0  30.0000  woman   
888         0       3  female   NaN      1      2  23.4500  woman   
889         1       1    male  26.0      0      0  30.0000    man   
890         0       3    male  32.0      0      0   7.7500    man   

     embark_town alone  
0    Southampton    no  
1      Cherbourg    no  
2    Southampton   yes  
3    Southampton    no  
4    Southampton   yes  
..           ...   ...  
886  Southampton   yes  
887  Southampton   yes  
888  Southampton    no  
889    Cherbourg   yes  
890   Queenstown   yes  

[891 rows x 10 columns]

In [12]:
#----------------------------------------------------------
# Información de las variables de la base de datos
#----------------------------------------------------------
datos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   who          891 non-null    object 
 8   embark_town  889 non-null    object 
 9   alone        891 non-null    object 
dtypes: float64(2), int64(4), object(4)
memory usage: 69.7+ KB


In [13]:
#----------------------------------------------------------
# Variables categóricas (tipo object)
#----------------------------------------------------------
datos.select_dtypes(include=['object']).describe()

sex  who  embark_town alone
count    891  891          889   891
unique     2    3            3     2
top     male  man  Southampton   yes
freq     577  537          644   537

Se hace un resumen de las variables cuantitativas (tipo entero), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

In [14]:
#----------------------------------------------------------
# Variables cuantitativas (tipo entero)
#----------------------------------------------------------
datos.select_dtypes(include=['int64']).describe()

survived      pclass       sibsp       parch
count  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642    0.523008    0.381594
std      0.486592    0.836071    1.102743    0.806057
min      0.000000    1.000000    0.000000    0.000000
25%      0.000000    2.000000    0.000000    0.000000
50%      0.000000    3.000000    0.000000    0.000000
75%      1.000000    3.000000    1.000000    0.000000
max      1.000000    3.000000    8.000000    6.000000

Se hace un resumen de las variables cuantitativas (tipo real), para analizar la media, la desviación estándar, mínimo, 25%, 50% (mediana), 75% y el máximo

In [15]:
#----------------------------------------------------------
# Variables cuantitativas (tipo real)
#----------------------------------------------------------
datos.select_dtypes(include=['float64']).describe()

age        fare
count  714.000000  891.000000
mean    29.699118   32.204208
std     14.526497   49.693429
min      0.420000    0.000000
25%     20.125000    7.910400
50%     28.000000   14.454200
75%     38.000000   31.000000
max     80.000000  512.329200

## 3. Seleccionar aleatoriamente el 80% del conjunto de datos para entrenar y el 20% restante para las pruebas

In [33]:
#----------------------------------------------------------
# Libreria sklearn
#----------------------------------------------------------
from sklearn.model_selection import train_test_split 

N=len(datos) # cantidad de datos
cTrain=int(N*0.8) # 80% para entrenar y 20% para probar
cTest=N-cTrain 

print(N,cTrain,cTest)# cantidad de: datos completos, datos entrenados, datos para probar

train_data,test_data= sklearn.model_selection.train_test_split(datos, train_size=cTrain, test_size=cTest)

891 712 179


891 es el número de observaciones en la base de datos

712 es el 80% de las observaciones de la base de datos que se van a utilizar para entrenar los datos

179 es el 20% de las observaciones de la base de datos que se van a utilizar para probar el modelo de predicción

In [17]:
#----------------------------------------------------------
# Dimensiones de los datos de entrenamiento
#----------------------------------------------------------
train_data.shape

(712, 10)

In [18]:
#----------------------------------------------------------
# Mostrar las primeras 5 filas de los datos de entrenamiento
#----------------------------------------------------------

train_data.head(5)

survived  pclass     sex   age  sibsp  parch     fare    who  \
639         0       3    male   NaN      1      0  16.1000    man   
173         0       3    male  21.0      0      0   7.9250    man   
872         0       1    male  33.0      0      0   5.0000    man   
409         0       3  female   NaN      3      1  25.4667  woman   
422         0       3    male  29.0      0      0   7.8750    man   

     embark_town alone  
639  Southampton    no  
173  Southampton   yes  
872  Southampton   yes  
409  Southampton    no  
422  Southampton   yes

In [19]:
#----------------------------------------------------------
# Estadísticas descriptivas datos de entrenamiento
#----------------------------------------------------------
train_data.describe()

survived      pclass         age       sibsp      parch        fare
count  712.000000  712.000000  566.000000  712.000000  712.00000  712.000000
mean     0.386236    2.342697   29.096290    0.523876    0.36236   31.576386
std      0.487228    0.830486   14.122773    1.061799    0.77633   47.962907
min      0.000000    1.000000    0.420000    0.000000    0.00000    0.000000
25%      0.000000    2.000000   20.000000    0.000000    0.00000    7.895800
50%      0.000000    3.000000   28.000000    0.000000    0.00000   14.400000
75%      1.000000    3.000000   37.000000    1.000000    0.00000   31.000000
max      1.000000    3.000000   74.000000    8.000000    6.00000  512.329200

## 4. Utilizar una estrategia para normalizar los datos



### Pipeline para los atributos categóricos

In [20]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.preprocessing import OneHotEncoder 
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

# 5 variables categoricas
cat_attribs = ['pclass','sex','who','embark_town','alone']

cat_pipeline = Pipeline([
("imputer", SimpleImputer(strategy="most_frequent")), 
("cat_encoder", OneHotEncoder(sparse=False))
])



### Pipeline para los atributos numéricos

In [32]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler 
from sklearn.preprocessing import MinMaxScaler

# Variables númericas

num_attribs = ['pclass','age', 'fare', 'sibsp', 'parch']
num_pipeline = Pipeline([("imputer", SimpleImputer(strategy="median")),
                         ("scaler", StandardScaler())])

### Pipeline completo

In [22]:
#----------------------------------------------------------
#Libreria sklearn
#----------------------------------------------------------
from sklearn.compose import ColumnTransformer

full_pipeline = ColumnTransformer([ ("num", num_pipeline, num_attribs), 
                                   ("cat", cat_pipeline, cat_attribs),])


In [29]:
#----------------------------------------------------------
# Se transforman los datos entrenados 
# Variables independientes
#----------------------------------------------------------

X_train = full_pipeline.fit_transform(train_data)
X_train.shape


(712, 18)

In [24]:
X_train[0,:]

array([ 0.79202445, -0.06922915, -0.32290091,  0.44872725, -0.46708788,
        0.        ,  0.        ,  1.        ,  0.        ,  1.        ,
        0.        ,  1.        ,  0.        ,  0.        ,  0.        ,
        1.        ,  1.        ,  0.        ])

In [30]:
#----------------------------------------------------------
# Variable dependiente -> survived
#----------------------------------------------------------

y_train = train_data["survived"]
y_train

639    0
173    0
872    0
409    0
422    0
      ..
431    1
832    0
539    1
834    0
258    1
Name: survived, Length: 712, dtype: int64

In [34]:
X_test = full_pipeline.transform(test_data) 
X_test

array([[ 0.79202445, -2.26012358, -0.48111965, ...,  0.        ,
         1.        ,  0.        ],
       [-0.41293582, -0.38698033, -0.1163463 , ...,  1.        ,
         1.        ,  0.        ],
       [-0.41293582,  0.08964645, -0.22066681, ...,  1.        ,
         1.        ,  0.        ],
       ...,
       [ 0.79202445, -0.86360711, -0.48564091, ...,  1.        ,
         0.        ,  1.        ],
       [ 0.79202445, -0.06922915, -0.34072302, ...,  0.        ,
         1.        ,  0.        ],
       [-0.41293582,  0.40739763, -0.22066681, ...,  1.        ,
         1.        ,  0.        ]])

In [36]:
y_test = test_data["survived"]
y_test

803    1
316    1
726    1
248    1
294    0
      ..
512    1
608    1
834    0
709    1
476    0
Name: survived, Length: 179, dtype: int64

## 5. Construir 5 redes neuronales variando en la topología de la red la cantidad de capas ocultas y de neuronas por cada capa oculta. Puede también variar los hiperparámetros solver y la función de activación. En todas las pruebas debe usar un random_state=123. Incluya en el notebook una tabla a manera de resumen con el accuracy obtenido en cada caso y también las matrices de confusión

In [50]:
#----------------------------------------------------------
# Librerias
#----------------------------------------------------------

from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

# Se crea una lista para almacenar los resultados
resultados = []

# Define una lista de topologías a probar
topologias = [
    (10,), # una capa oculta con 10 neuronas
    (10, 10), # dos capas ocultas con 10 neuronas cada una
    (20,), # una capa oculta con 20 neuronas
    (20, 20), # dos capas ocultas con 20 neuronas cada una
    (30,), # una capa oculta con 30 neuronas
    ]

# Define una lista de funciones de activación a probar
func_activacion = ["relu", "logistic"] 

# Define una lista de solvers a probar
solvers = ["adam", "sgd"]

# Itera para cada topología
for topologia in topologias:

    # Itera para cada función de activación
    for activacion in func_activacion:
    
        # Itera para cada solver
        for solver in solvers:

            # Crea el modelo de red neuronal
            model = MLPClassifier(hidden_layer_sizes=topologia, activation=activacion, solver=solver, random_state=123)
            
            # Entrena el modelo
            model.fit(X_train, y_train)
            
            # Realiza las predicciones
            y_pred = model.predict(X_test)
            
            # Calcula el accuracy
            accuracy = round(accuracy_score(y_test, y_pred),4) # con 4 decimales
            
            # Calcula la matriz de confusión
            confusion = confusion_matrix(y_test, y_pred)
            
            # Almacena el resultado
            resultados.append({"Topología": topologia, "Función de activación": activacion, "Solver": solver, "Accuracy": accuracy, "Matriz de confusión": confusion})

# Imprime los resultados
print(resultados)




/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:

[{'Topología': (10,), 'Función de activación': 'relu', 'Solver': 'adam', 'Accuracy': 0.7933, 'Matriz de confusión': array([[95, 12],
       [25, 47]])}, {'Topología': (10,), 'Función de activación': 'relu', 'Solver': 'sgd', 'Accuracy': 0.7989, 'Matriz de confusión': array([[96, 11],
       [25, 47]])}, {'Topología': (10,), 'Función de activación': 'logistic', 'Solver': 'adam', 'Accuracy': 0.8045, 'Matriz de confusión': array([[96, 11],
       [24, 48]])}, {'Topología': (10,), 'Función de activación': 'logistic', 'Solver': 'sgd', 'Accuracy': 0.5978, 'Matriz de confusión': array([[107,   0],
       [ 72,   0]])}, {'Topología': (10, 10), 'Función de activación': 'relu', 'Solver': 'adam', 'Accuracy': 0.8101, 'Matriz de confusión': array([[97, 10],
       [24, 48]])}, {'Topología': (10, 10), 'Función de activación': 'relu', 'Solver': 'sgd', 'Accuracy': 0.7989, 'Matriz de confusión': array([[98,  9],
       [27, 45]])}, {'Topología': (10, 10), 'Función de activación': 'logistic', 'Solver': '

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


In [56]:
#--------------------------------------------------------------------------
# Función para imprimir en una tabla los resultados de las redes neuronales
#--------------------------------------------------------------------------

def imprimir_Tabla(datos):
    # Imprime la primera fila de la tabla
    print("|","Topología", "|","Función de activación", "|","solver", "|","accuracy","|" ,"Matriz de confusion")
    
    # Imprime la línea de separación
    print("+------- "*len(datos[0]) + "+")
    
    # Itera para cada resultado
    for resultado in datos:
        # Crea una fila con los datos del resultado
        fila = [resultado["Topología"], resultado["Función de activación"], resultado["Solver"], resultado["Accuracy"], resultado["Matriz de confusión"]]
        
        # Itera para cada elemento de la fila
        for elemento in fila:
            # Imprime el elemento
            print("| ",str(elemento),"", end="")
        print("|")
    
    # Imprime la línea de separación final
    print("+------- "*len(datos[0]) + "+")

# Imprime la tabla
imprimir_Tabla(resultados)



| Topología | Función de activación | solver | accuracy | Matriz de confusion
+------- +------- +------- +------- +------- +
|  (10,) |  relu |  adam |  0.7933 |  [[95 12]
 [25 47]] |
|  (10,) |  relu |  sgd |  0.7989 |  [[96 11]
 [25 47]] |
|  (10,) |  logistic |  adam |  0.8045 |  [[96 11]
 [24 48]] |
|  (10,) |  logistic |  sgd |  0.5978 |  [[107   0]
 [ 72   0]] |
|  (10, 10) |  relu |  adam |  0.8101 |  [[97 10]
 [24 48]] |
|  (10, 10) |  relu |  sgd |  0.7989 |  [[98  9]
 [27 45]] |
|  (10, 10) |  logistic |  adam |  0.8101 |  [[97 10]
 [24 48]] |
|  (10, 10) |  logistic |  sgd |  0.5978 |  [[107   0]
 [ 72   0]] |
|  (20,) |  relu |  adam |  0.7933 |  [[95 12]
 [25 47]] |
|  (20,) |  relu |  sgd |  0.8101 |  [[98  9]
 [25 47]] |
|  (20,) |  logistic |  adam |  0.8045 |  [[96 11]
 [24 48]] |
|  (20,) |  logistic |  sgd |  0.6089 |  [[107   0]
 [ 70   2]] |
|  (20, 20) |  relu |  adam |  0.8101 |  [[94 13]
 [21 51]] |
|  (20, 20) |  relu |  sgd |  0.7877 |  [[96 11]
 [27 45]] |
| 

### 6. Indique en el notebook usando una celda de tipo Markdown los hiperparámetros que por el momento le permiten obtener la red con mayor accuracy

In [64]:
# Encontrar la red neuronal con el mayor valor de Accuracy
max_accuracy = max(resultados, key=lambda x: x["Accuracy"])

# Imprime los hiperparámetros del mejor resultado
print("-------------------------------------------")
print("Mejores hiperparámetros:")
print("-------------------------------------------")
print("Topología:", max_accuracy["Topología"])
print("Función de activación:", max_accuracy["Función de activación"])
print("Solver:", max_accuracy["Solver"])
print("Accuracy:", max_accuracy["Accuracy"])
print("Matriz de confusión:", max_accuracy["Matriz de confusión"])

-------------------------------------------
Mejores hiperparámetros:
-------------------------------------------
Topología: (30,)
Función de activación: relu
Solver: adam
Accuracy: 0.8156
Matriz de confusión: [[100   7]
 [ 26  46]]


### 7. Seleccione uno de los hiperparámetros disponibles en la documentación (https://scikit-learn.org/stable/modules/generated/sklearn.neural_network.MLPClassifier.html) que sea diferente al solver, a la función de activación, y al random_state. Realice dos variaciones en el hiperparámetro seleccionado manteniendo los otros hiperparámetros del punto anterior. Indique el accuracy obtenido al modificar el hiperparámetro seleccionado y analice si la red mejora, empeora, o mantiene su exactitud. Incluya en el notebook dicho análisis




In [69]:
#-----------------------------------------------
# Librerias
#-----------------------------------------------
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import accuracy_score

# Se crea una lista para almacenar los resultados
resultados = []

# Define una lista de tasas de aprendizaje inicial para la prueba
learning_rate_inits = [0.01, 0.001]

# Itera para cada tasa de aprendizaje inicial
for learning_rate_init in learning_rate_inits:

    # Define la topología de la red
    hidden_layer_sizes = (10, 10) # dos capas ocultas con 10 neuronas cada una
    activation = "relu" # función de activación ReLU
    solver = "adam" # solver Adam
    
    # Crea el modelo de red neuronal
    model = MLPClassifier(hidden_layer_sizes=hidden_layer_sizes, activation=activation, solver=solver, random_state=123, learning_rate_init=learning_rate_init)
    
    # Entrena el modelo
    model.fit(X_train, y_train)
    
    # Realiza las predicciones
    y_pred = model.predict(X_test)
    
    # Calcula el accuracy
    accuracy = accuracy_score(y_test, y_pred)
    
    # Almacena el resultado
    resultados.append({"learning_rate_init": learning_rate_init, "accuracy": accuracy})

# Imprime los resultados
print("---------------------------------")
print(resultados)
print("---------------------------------")

/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(


[{'learning_rate_init': 0.01, 'accuracy': 0.8268156424581006}, {'learning_rate_init': 0.001, 'accuracy': 0.8100558659217877}]
---------------------------------


/usr/local/lib/python3.8/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:692: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  warnings.warn(
